### PR TITLE
[MILO MIGRATION] feat: port decorateLinks feature from milo

### PR DESCRIPTION
**Pull request by catalan-adobe**
Port Milo's link decoration into Express.

Resolves: [MWPW-139145](https://jira.corp.adobe.com/browse/MWPW-139145), [MWPW-137709](https://jira.corp.adobe.com/browse/MWPW-137709)

From the changes, we detected 3 express custom features that are not in Milo's `decorateLinks` code:
* enforce same window navigation for a defined set of domains
* handle "contact links" (`tel:`, `mailto:`, `sms:`)
* handle link replacement on sheet-powered pages

Resolution for these 3 points are tracked in related sub tasks of the main ticket.

Test URLs:
- General link features:
  - Before: https://main--express--adobecom.hlx.page/drafts/catalan/decoratelinks-milo-port?lighthouse=on
  - After: https://decoratelinks-milo--express--adobecom.hlx.page/drafts/catalan/decoratelinks-milo-port?lighthouse=on
- spreadsheet replacement:
  - Before: https://main--express--adobecom.hlx.page/express/templates/curriculum-vitae?lighthouse=on
  - After: https://decoratelinks-milo--express--adobecom.hlx.page/express/templates/curriculum-vitae?lighthouse=on
- modal fragment link replacement:
  - Before: https://main--express--adobecom.hlx.page/express/business?lighthouse=on#sales-contact-form-1
  - After: https://decoratelinks-milo--express--adobecom.hlx.page/express/business?lighthouse=on#sales-contact-form-1

### CODE REVIEW

File: express/blocks/search-marquee/search-marquee.js [hunk 1]
```diff
@@ -251,7 +251,7 @@ async function decorateBackground(block) {
   const mediaRow = block.querySelector('div:nth-child(2)');
   return new Promise((resolve) => {
     if (mediaRow) {
-      const media = mediaRow.querySelector('a')?.href || mediaRow.querySelector(':scope > div')?.textContent;
+      const media = mediaRow.querySelector('picture img')?.src || mediaRow.querySelector(':scope > div')?.textContent;
       mediaRow.remove();
       if (media) {
         const splitArr = media.split('.');
```
Comment (vhargrave) [inline express/blocks/search-marquee/search-marquee.js:257]: We only need half this code now. We already have the image in an image tag. Let's just give it the right class, move it to where it needs to be, make it load eagerly and await that.

**Review by vhargrave — CHANGES_REQUESTED** [review]
We're getting closer! 
found a few small things, and one more bug. Look at the contact-sales button [here](https://decoratelinks-milo--express--adobecom.hlx.page/express/business) . It's a link, but on main you'll see that it's actually supposed to be a button

File: express/scripts/content-replace.js [hunk 2]
```diff
@@ -125,6 +125,19 @@ async function autoUpdatePage(main) {
     }
     return match;
   });
+
+  // handle link replacement on sheet-powered pages
+  main.querySelectorAll('a[href*="#"]').forEach((a) => {
```
[the remark this answers is not in the document]
Comment (catalan-adobe) [reply express/scripts/content-replace.js:130]: Why would you move it to scripts.js?
About `isDoc`, it's defined in utils.js but not scripts.js, what about moving this code in the `if (isDoc) {` block of utils.js? (https://github.com/adobecom/express/blob/decoratelinks-milo/express/scripts/utils.js#L2607, few lines above its current location)

File: express/blocks/search-marquee/search-marquee.js (hunk 1 above)
Comment (catalan-adobe) [reply express/blocks/search-marquee/search-marquee.js]: fixed in [d0ae765](https://github.com/adobecom/express/pull/484/commits/d0ae7657ea253ec030fa74875e04487b9dd9cb13)

File: express/scripts/content-replace.js [hunk 3]
```diff
@@ -125,6 +125,19 @@ async function autoUpdatePage(main) {
...
+    try {
+      let url = new URL(a.href);
+      if (getMetadata(url.hash.replace('#', ''))) {
+        a.href = getMetadata(url.hash.replace('#', ''));
+        url = new URL(a.href);
+      }
+    } catch (e) {
+      // invalid url
```
[the remark this answers is not in the document]
Comment (catalan-adobe) [reply express/scripts/content-replace.js]: fixed in [649f15b](https://github.com/adobecom/express/pull/484/commits/649f15b594735a406ff40ff15f8fd24e50f3f520)

File: express/scripts/utils.js [hunk 4]
```diff
@@ -714,83 +851,76 @@ export function decorateAutoBlock(a) {
...
 
       // Modals
-      if (url.hash !== '' && !isInlineFrag && !videoTag) {
+      if (url.hash !== '' && !isInlineFrag) {
         a.dataset.modalPath = url.pathname;
         a.dataset.modalHash = url.hash;
         a.href = url.hash;
-        a.className = 'modal';
-        a.setAttribute('data-block-name', 'modal');
+        a.className = `modal link-block ${[...a.classList].join(' ')}`;
```
Comment (catalan-adobe) [inline express/scripts/utils.js:891]: this is the "culprit" line for the non rendering secondary button on https://decoratelinks-milo--express--adobecom.hlx.page/express/business.
Removing `link-block` fixes the issue.
@vhargrave, `link-block` usage is limited to `utils.js`. From what I understand, it is used as a way to exclude links from being decorated via `decorateButtons`.
I wonder if we could simply remove `link-block` everywhere.

Comment (catalan-adobe) [reply express/scripts/utils.js:891]: @vhargrave, I applied your suggestion in [77ed2ef](https://github.com/adobecom/express/pull/484/commits/77ed2ef26fbd5208744407770f169a2a34a397d8)

**Review by vhargrave — APPROVED** [review]
It's looking good! Awesome stuff and thank you so much for the help on this :-)

File: express/blocks/search-marquee/search-marquee.js [hunk 5]
```diff
@@ -246,31 +246,21 @@ async function decorateSearchFunctions(block) {
...
-          } else {
-            block.prepend(backgroundImg);
-          }
-          backgroundImg.onload = () => resolve();
-        } else {
-          resolve();
-        }
+  if (mediaRow) {
+    const media = mediaRow.querySelector('picture img');
+    mediaRow.remove();
```
[the remark this answers is not in the document]
Comment (catalan-adobe) [reply express/blocks/search-marquee/search-marquee.js]: done in [63b8116](https://github.com/adobecom/express/pull/484/commits/63b811647001b4390316a3a981eec00a40b87a8b)

**Review by apganapa — APPROVED** [review]

**Pull request merged**